pkg/trafficrouting/manager.go

```diff
@@ -52,142 +70,144 @@ func NewTrafficRoutingManager(c client.Client) *Manager {
 	stableService := &corev1.Service{}
-	err := m.Get(context.TODO(), client.ObjectKey{Namespace: c.Rollout.Namespace, Name: trafficRouting.Service}, stableService)
+	err := m.Get(context.TODO(), client.ObjectKey{Namespace: c.Namespace, Name: trafficRouting.Service}, stableService)
 	if err != nil {
-		klog.Errorf("rollout(%s/%s) get stable service(%s) failed: %s", c.Rollout.Namespace, c.Rollout.Name, trafficRouting.Service, err.Error())
+		klog.Errorf("%s get stable service(%s) failed: %s", c.Key, trafficRouting.Service, err.Error())
 		// not found, wait a moment, retry
 		if errors.IsNotFound(err) {
 			return false, nil
 		}
 		return false, err
 	}
 	// canary service name
-	canaryServiceName := fmt.Sprintf("%s-canary", trafficRouting.Service)
-	// fetch canary service
+	canaryServiceName := getCanaryServiceName(trafficRouting.Service, c.OnlyTrafficRouting)
 	canaryService := &corev1.Service{}
-	err = m.Get(context.TODO(), client.ObjectKey{Namespace: c.Rollout.Namespace, Name: canaryServiceName}, canaryService)
-	if err != nil && !errors.IsNotFound(err) {
-		klog.Errorf("rollout(%s/%s) get canary service(%s) failed: %s", c.Rollout.Namespace, c.Rollout.Name, canaryServiceName, err.Error())
-		return false, err
-	} else if errors.IsNotFound(err) {
-		canaryService, err = m.createCanaryService(c, canaryServiceName, *stableService.Spec.DeepCopy())
-		if err != nil {
+	canaryService.Namespace = stableService.Namespace
+	canaryService.Name = canaryServiceName
+	// In full-link grayscale scenario(a -> b -> c), if only b or c is released,
+	//and a is not released in this scenario, then the canary service is not needed.
+	if !c.OnlyTrafficRouting {
+		if c.StableRevision == "" || c.CanaryRevision == "" {
+			klog.Warningf("%s stableRevision or podTemplateHash can not be empty, and wait a moment", c.Key)
+			return false, nil
+		}
+		// fetch canary service
+		err = m.Get(context.TODO(), client.ObjectKey{Namespace: c.Namespace, Name: canaryServiceName}, canaryService)
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("%s get canary service(%s) failed: %s", c.Key, canaryServiceName, err.Error())
 			return false, err
+		} else if errors.IsNotFound(err) {
+			canaryService, err = m.createCanaryService(c, canaryServiceName, *stableService.Spec.DeepCopy())
+			if err != nil {
+				return false, err
+			}
 		}
-	}
 
-	// patch canary service only selector the canary pods
-	if canaryService.Spec.Selector[c.Workload.RevisionLabelKey] != canaryStatus.PodTemplateHash {
-		body := fmt.Sprintf(`{"spec":{"selector":{"%s":"%s"}}}`, c.Workload.RevisionLabelKey, canaryStatus.PodTemplateHash)
-		if err = m.Patch(context.TODO(), canaryService, client.RawPatch(types.StrategicMergePatchType, []byte(body))); err != nil {
-			klog.Errorf("rollout(%s/%s) patch canary service(%s) selector failed: %s", c.Rollout.Namespace, c.Rollout.Name, canaryService.Name, err.Error())
-			return false, err
+		// patch canary service only selector the canary pods
```

Comment:
patch canary service only selector the canary pods -> patch canary service to only select canary pods